entity-framework/core/modeling/value-conversions.md

```diff
@@ -2,145 +2,746 @@
+        {
+            [JsonConstructor]
+            public AnnualFinance(int year, Money income, Money expenses)
+            {
+                Year = year;
+                Income = income;
+                Expenses = expenses;
+            }
+
+            public int Year { get; }
+            public Money Income { get; }
+            public Money Expenses { get; }
+            public Money Revenue => new Money(Income.Amount - Expenses.Amount, Income.Currency);
+        }
+-->
+[!code-csharp[ValueObjectCollection](../../../samples/core/Modeling/ValueConversions/ValueObjectCollection.cs?name=ValueObjectCollection)]
+
+This type composes several of the `Money` types we created previously:
+
+<!--
+        public readonly struct Money
+        {
+            [JsonConstructor]
+            public Money(decimal amount, Currency currency)
+            {
+                Amount = amount;
+                Currency = currency;
+            }
+
+            public override string ToString()
+                => (Currency == Currency.UsDollars ? "$" : "£") + Amount;
+
+            public decimal Amount { get; }
+            public Currency Currency { get; }
+        }
+
+        public enum Currency
+        {
+            UsDollars,
+            PoundsStirling
+        }
+-->
+[!code-csharp[ValueObjectCollectionMoney](../../../samples/core/Modeling/ValueConversions/ValueObjectCollection.cs?name=ValueObjectCollectionMoney)]
+
+We can then add a collection of `AnnualFinance` to our entity type:
+
+<!--
+        public class Blog
+        {
+            public int Id { get; set; }
+            public string Name { get; set; }
+            
+            public IList<AnnualFinance> Finances { get; set; }
+        }
+-->
+[!code-csharp[ValueObjectCollectionModel](../../../samples/core/Modeling/ValueConversions/ValueObjectCollection.cs?name=ValueObjectCollectionModel)]
+
+And again use serialization to store this:
+
+<!--
+                modelBuilder.Entity<Blog>()
+                    .Property(e => e.Finances)
+                    .HasConversion(
+                        v => JsonSerializer.Serialize(v, null),
+                        v => JsonSerializer.Deserialize<List<AnnualFinance>>(v, null),
+                        new ValueComparer<IList<AnnualFinance>>(
+                            (c1, c2) => c1.SequenceEqual(c2),
+                            c => c.Aggregate(0, (a, v) => HashCode.Combine(a, v.GetHashCode())),
+                            c => (IList<AnnualFinance>)c.ToList()));
+-->
+[!code-csharp[ConfigureValueObjectCollection](../../../samples/core/Modeling/ValueConversions/ValueObjectCollection.cs?name=ConfigureValueObjectCollection)]
 
-```csharp
-public class Rider
-{
-    public int Id { get; set; }
+> [!NOTE]
+> As before, this conversion requires a <xref:Microsoft.EntityFrameworkCore.ChangeTracking.ValueComparer%601>. See [Value Comparers](xref:core/modeling/value-comparers) for more information.
+
+### Value objects as keys
```

Comment:
Do we really want to show this, given that I'm not sure we like it? :smiling_imp:

Reply:
Yes, because so many people ask about this, and many of them get it wrong.